### PR TITLE
https://github.com/AdguardTeam/AdguardBrowserExtension/issues/463

### PR DESCRIPTION
**Pull request by atropnikov**

### CODE REVIEW

File: Extension/lib/content-script/preload.js [hunk 1]
```diff
@@ -237,6 +237,91 @@
...
     /**
+     * This function prepares calculation of css hits.
+     * We are waiting for DOMContentLoaded and start calculation.
+     */
+    function prepareCssHitsCounter() {
+        // Document can already be loaded
+        if (['interactive', 'complete'].indexOf(document.readyState) >= 0) {
+            countCssHits();
+        } else {
+            addEventListener('DOMContentLoaded', countCssHits);
```
Comment (Mizzick) [inline Extension/lib/content-script/preload.js]: Are you sure all the styles gonna be applied by this moment?

File: Extension/lib/content-script/preload.js [hunk 2]
```diff
@@ -288,6 +373,10 @@
         if (response && response.selectors && response.selectors.css && response.selectors.css.length > 0) {
             addIframeHidingStyle();
         }
+
+        if (response && response.selectors && response.selectors.cssHitsCounterEnabled) {
+            prepareCssHitsCounter();
```
Comment (Mizzick) [inline Extension/lib/content-script/preload.js]: I think this should take the place on styles applied callback or smth

File: Extension/lib/content-script/preload.js [hunk 3]
```diff
@@ -237,6 +237,91 @@
...
+     * 2. For each element from sub collection: retrieve calculated css 'content' attribute and if it contains 'adguard' marker then retrieve rule text and filter identifier.
+     * 3. Start next task with some delay.
+     *
+     * @param elements Collection of all elements
+     * @param start Start of batch
+     * @param end End of batch
+     * @param result Collection for save result
+     * @param callback Finish callback
+     */
+    function countCssHitsBatch(elements, start, end, result, callback) {
```
Comment (Mizzick) [inline Extension/lib/content-script/preload.js]: Extract to some utils? And also would be nice to have tests for this stuff.

File: Extension/tests/css-filter/test-css-filter.js [hunk 4]
```diff
@@ -519,35 +483,35 @@ QUnit.test("Extended Css Build CssHits", function (assert) {
...
     css = selectors.css;
     extendedCss = selectors.extendedCss;
     commonCss = filter.buildCssHits(null).css;
     assert.equal(commonCss.length, 1);
-    assert.equal(commonCss[0].trim(), shadowDomPrefix + ".banner { display: none!important; background-image: url('hit-prefix/elemhidehit.png#%3B%23%23.banner') !important;}");
+    assert.equal(commonCss[0].trim(), shadowDomPrefix + ".banner { display: none!important; content: 'adguard2%3B%23%23.banner' !important;}");
     assert.equal(css.length, 2);
-    assert.equal(css[0].trim(), shadowDomPrefix + ".banner { display: none!important; background-image: url('hit-prefix/elemhidehit.png#%3B%23%23.banner') !important;}");
-    assert.equal(css[1].trim(), shadowDomPrefix + ".sponsored { display: none!important; background-image: url('hit-prefix/elemhidehit.png#%3Badguard.com%23%23.sponsored') !important;}");
+    assert.equal(css[0].trim(), shadowDomPrefix + ".banner { display: none!important; content: 'adguard2%3B%23%23.banner' !important;}");
```
Comment (Mizzick) [inline Extension/tests/css-filter/test-css-filter.js]: the template could be more elegant, something like 
`adguard-css-hits-filter-[filterId]-rule-[ruleText]`

File: Extension/lib/content-script/preload.js (hunk 1 above)
Comment (atropnikov) [inline Extension/lib/content-script/preload.js]: Yes. This method is invoked after ```applyCss```.

File: Extension/tests/css-filter/test-css-filter.js (hunk 4 above)
Comment (atropnikov) [inline Extension/tests/css-filter/test-css-filter.js]: I think it doesn't matter)

File: Extension/lib/content-script/preload.js (hunk 1 above)
Comment (atropnikov) [inline Extension/lib/content-script/preload.js]: Fixed. Now computation starts after 'load' event with 2 seconds delay.

File: Extension/lib/content-script/preload.js (hunk 3 above)
Comment (atropnikov) [inline Extension/lib/content-script/preload.js]: Moved. I'll add tests later.

File: Extension/lib/utils/css-hits-counter.js [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+            });
+        }
+
+        if (length === elements.length) {
+            callback(result);
+            return;
+        }
+
+        start = end;
+        end += 100;
```
Comment (Mizzick) [inline Extension/lib/utils/css-hits-counter.js]: 100 could be represented as parameter or through end-start

File: Extension/browser/firefox/lib/contentScripts.js [hunk 6]
```diff
@@ -194,6 +198,27 @@
     }
 
     /**
+     * Register css in user sheet
+     * @param cssContent Css content
+     */
+    function registerCss(cssContent) {
```
Comment (ameshkov) [inline Extension/browser/firefox/lib/contentScripts.js]: Why? Can't we get rid of it completely?

File: Extension/tests/css-filter/test-css-filter.js [hunk 7]
```diff
@@ -519,35 +483,35 @@ QUnit.test("Extended Css Build CssHits", function (assert) {
...
+    assert.equal(css[1].trim(), shadowDomPrefix + ".sponsored { display: none!important; content: 'adguard1%3Badguard.com%23%23.sponsored' !important;}");
     assert.equal(extendedCss.length, 1);
-    assert.equal(extendedCss[0].trim(), ".sponsored[-ext-contains=test] { display: none!important; background-image: url('hit-prefix/elemhidehit.png#%3Badguard.com%23%23.sponsored%5B-ext-contains%3Dtest%5D') !important;}");
+    assert.equal(extendedCss[0].trim(), ".sponsored[-ext-contains=test] { display: none!important; content: 'adguard1%3Badguard.com%23%23.sponsored%5B-ext-contains%3Dtest%5D' !important;}");
 
-    selectors = filter.buildCssHits("adguard.com", 'hit-prefix', true);
+    selectors = filter.buildCssHits("adguard.com", true);
```
Comment (ameshkov) [inline Extension/tests/css-filter/test-css-filter.js]: I suppose we need a way of unit-testing the whole feature:

1. Building proper CSS with hits counter
2. Collecting rule hits by the content script
3. Processing of these stats by the CssHitsCounter

File: Extension/browser/firefox/lib/contentScripts.js (hunk 6 above)
Comment (atropnikov) [inline Extension/browser/firefox/lib/contentScripts.js]: No, we can't. Chrome and Safari allow to register css via manifest file, but Firefox doesn't. We need to register assistant's selector style.

Comment (ameshkov) [inline Extension/browser/firefox/lib/contentScripts.js]: Ah, got it

**Pull request merged**